### PR TITLE
Montana Child Tax Credit

### LINKED ISSUE #2326: Montana Child Tax Credit
https://leg.mt.gov/bills/2023/billhtml/HB0268.htm

### PR DESCRIPTION
**Pull request by Yoon-Eric**
### Summary
🧒💵🏔️

This pull request adds a new variable and parameters for the Montana Child Tax Credit (MT CTC), a state-level tax credit for families with children. The variable `ctc.py` defines the logic and attributes of the credit, and the parameter files in the `ctc` folder specify the eligibility and reduction rules for the credit.

> _`MT CTC` class_
> _New parameters for income_
> _Autumn tax season_

### Walkthrough
*  Add a new variable class for the Montana Child Tax Credit (MT CTC) in `ctc.py` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-ea8e37de4510af8a89f189a32d85a72c4098f112463deeaf49680fed260a3369L1-R17))
*  Add parameter files for the MT CTC eligibility and amount in the same folder ([link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-5543ae3552cefecab569e2446f039ad47edacec6625988712836a6b9d2608344L1-R9), [link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-eaf3abd8da41cc1c9932529282bab8147bd7ee18985465b137d0bb68fb96d763L1-R9), [link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-ba65ed7db9611143053944ca6c4a3afcc41556aac5a295128cba09f23ddda8b4L1-R9), [link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-ae2a752a17be53826a7c01c3187ceb097f9898196e8f6cab4d569981e398e8daL1-R9))
*  Remove the redundant parameter file for the investment income threshold in the same folder ([link](https://github.com/PolicyEngine/policyengine-us/pull/2328/files?diff=unified&w=0#diff-48318e896e028c1c616fa99876e608aaa403e5239e4236a56131eb4d5515f57dL1-R9))

fixes #2326

### CODE REVIEW

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+description: Child Tax Credit Amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:1]:
```suggestion
description: Montana provides the following base amount per child under the child tax credit.
```

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+values:
+  2023-01-01: 1_200
+metadata:
+  units: currency-USD
+  period: year
+  label: MT child tax credit amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:7]:
```suggestion
  label: Montana child tax credit base
```

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:1]: rename file to "base.yaml"

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml (hunk 2 above)
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:7]: and in other parameters

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml (hunk 1 above)
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:1]: Similar description structure for other parameters

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  reference:
+    - title: MT
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:9]:
```suggestion
    - title: 2023 Montana Legislature, House Bill No. 268, Section 1, 7 (c)
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/amounts.yaml:9]: And other parameters

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/child_tax_credit.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            tax_unit("investment_income", period) <= p.investment_threshold
+        )
+        person = tax_unit.members
+        dependent = person("is_tax_unit_dependent", period)
+        meets_age_limit = person("age", period) < p.child_age_eligibility
+        eligible_child = dependent & meets_age_limit
+        eligible_children = tax_unit.sum(eligible_child)
+        
+        # reduction
+        if gross_income > p.reduction.treshold:
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/child_tax_credit.py:37]: we can not use "if" statements, instead we have to use`where(` statements

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/child_tax_credit.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            return income_eligible * investment_income_eligible * eligible_children * p.amount * p.reduction.rate * ((gross_income - p.reduction.treshold) // p.reduction.increment)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/child_tax_credit.py:38]: break this up into steps for readability

**Review by PavelMakarchuk — COMMENTED** [review]
comment

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        eligible_child = dependent & meets_age_limit
+
+        eligible_children = tax_unit.any(eligible_child)
+
+        eligible = income_eligible * investment_income_eligible
+
+        child_amount = eligible_children * p.base
+        credit = eligible * child_amount
+        # reduction
+        reduction_rate = p.reduction.rate * (
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py:34]: Break up for clarity
```suggestion
        excess = max_(gross_income - p.reduction.threshold, 0)
        increments = excess // p.reduction.increment
        reduction_rate = p.reduction.rate * increments
```

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            max_((gross_income - p.reduction.threshold), 0)
+            // p.reduction.increment
+        )
+        reduced_credit = max_((credit - reduction_rate), 0)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py:38]: We conventionally skip these unnecessary parens
```suggestion
        reduced_credit = max_(credit - reduction_rate, 0)
```

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+        return where(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py:40]: Is this `where` needed? Looks like it'll equal `reduced_credit`.

File: policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        income_eligible = gross_income <= p.income_threshold
+        investment_income_eligible = (
+            tax_unit("net_investment_income", period) <= p.investment_threshold
+        )
+        person = tax_unit.members
+        dependent = person("is_tax_unit_dependent", period)
+        meets_age_limit = person("age", period) < p.child_age_eligibility
+        eligible_child = dependent & meets_age_limit
+
+        eligible_children = tax_unit.any(eligible_child)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/mt/tax/income/credits/ctc/mt_ctc.py:27]: Do you mean `tax_unit.sum`?

Please add a test case with multiple eligible children to check.

If it is `tax_unit.any` (that is, if it's not scaled by number of kids), then please rename to `has_eligible_children`.

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+description: Montana provides the following base amount per child under the child tax credit.
+values:
+  2023-01-01: 1_200
+metadata:
+  units: currency-USD
+  period: year
+  label: Montana child tax credit base
+  reference:
+    - title: 2023 Montana Legislature, House Bill No. 268, Section 1, 7 (c)
+    - href: https://leg.mt.gov/bills/2023/billhtml/HB0268.htm
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml:10]:
```suggestion
      href: https://leg.mt.gov/bills/2023/billhtml/HB0268.htm
```

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+description: Montana provides the following base amount per child under the child tax credit.
+values:
+  2023-01-01: 1_200
+metadata:
+  units: currency-USD
+  period: year
+  label: Montana child tax credit base
+  reference:
+    - title: 2023 Montana Legislature, House Bill No. 268, Section 1, 7 (c)
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml:9]: please add a reference to the legal code (which differs from legislation)

also include references to tax forms

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml (hunk 11 above)
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml:9]: it seems like the bill has not been implemented yet, but the structure is fully in place now, should make this a reform PR?

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+description: Montana provides the following base amount per child under the child tax credit.
+values:
+  2023-01-01: 1_200
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/base.yaml:3]: all yaml files

File: policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/child_age_eligibility.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+description: Child Age For Child Tax Credit Eligibility.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/child_age_eligibility.yaml:1]:
```suggestion
description: Montana provides the child tax credit for filers with children at or below this age.
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/mt/tax/income/credits/ctc/child_age_eligibility.yaml:1]: all yaml files similar descriptions

**Pull request closed without merging**